### ISSUE TITLE
RequireThis: Incorrectly locates errors for final instance variable

### ISSUE THREAD

**Issue by anhminhtran235** [checkstyle/checkstyle#8973]
Check documentation: https://checkstyle.org/config_coding.html#RequireThis
Discovered in #8873
```
$ cat Test.java
public class Test {
  final int x = 0;
  public Test(int x) {
    x = x; // wrong error message
  }
}

$ cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis" />
    </module>
</module>

java -jar .\checkstyle-8.37-all.jar -c config.xml Test.java
Starting audit...
[ERROR] E:\checkstyle-playground\Test.java:5:5: Reference to instance variable 'x' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
```

Since `x` is a `final` instance variable, assigning `this.x = x` will give an error. Therefore, the error should be at 5:9, instead of 5:5 (which would make the code: `x = this.x`)

Similar problems happen for instance variable in `record` (which is `private final` by default) **in java 15**:
```
$ cat InputRequireThis.java
public record InputRequireThis(int x, int y) {
  InputRequireThis(int x) {
    this(x, 42);
    x = x; // wrong error message
  }
}

$ cat config.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis" />
    </module>
</module>

java -jar .\checkstyle-8.37-all.jar -c config.xml InputRequireThis.java
Starting audit...
[ERROR] E:\checkstyle-playground\InputRequireThis.java:4:5: Reference to instance variable 'x' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 1 errors.
```
The error should be at 4:9 instead of 4:5.

**Comment by anhminhtran235** [comment]
@romani Is this issue approved? Can I work on it?

**Comment by romani** [comment]
Approved

**Comment by anhminhtran235** [comment]
I am on it

**Comment by anhminhtran235** [comment]
@romani I need some help, please. There are many tricky cases related to final instance variables.
First of all, there are [3 places](https://stackoverflow.com/questions/18584674/can-a-final-variable-be-initialized-when-an-object-is-created) where we can init a final instance variable:
- in the constructor
- in an instance initialization block.
- when we declare it

As the result, there are edge cases such as:
```
public class Test {
  final int x;
  final int y = 1;
  final int z;

  public Test(int x, int y, int z) {
    this.x = x; // compiles fine, as instance variable x has not been init
    this.y = y; // compile error, y has been init
    this.z = z; // compile error, instance initialization block take precedence
  }

  {
    z = 2; // Is called before constructor
  }
}
```

Currently, here's what RequireThis will do:
```
$ cat Test.java
public class Test {
  final int x;
  final int y = 1;
  final int z;

  public Test(int x, int y, int z) {
    x = x;
    y = y;
    z = z;
  }

  {
    z = 2; // Is called before constructor
  }
}

$ java -jar .\checkstyle-8.37-all.jar -c config.xml Test.java
Starting audit...
[ERROR] E:\checkstyle-playground\Test.java:7:5: Reference to instance variable 'x' needs "this.". [RequireThis]
[ERROR] E:\checkstyle-playground\Test.java:8:5: Reference to instance variable 'y' needs "this.". [RequireThis]
[ERROR] E:\checkstyle-playground\Test.java:9:5: Reference to instance variable 'z' needs "this.". [RequireThis]
Audit done.
Checkstyle ends with 3 errors.
```
It correctly locates the error at 7:5, but incorrectly locates the errors at 8:5 and 9:5 (should be 8:9 and 9:9).

So **my first question is**: Should we add code to handle these 2 edge cases? I think this is such a rare case, and it does not worth it.

**Comment by anhminhtran235** [comment]
@romani In other places rather than constructor, RequireThis currently behaves differently:

```
$ cat Test.java
public class Test {
  final int x = 1;

  public void foo(int x) {
    x = x; // currently no violation
  }
}

$ java -jar .\checkstyle-8.37-all.jar -c config.xml Test.java
Starting audit...
Audit done.
```
There should be a violation at 5:9, but currently there isn't. My **second question is**: Should we add this violation? If we do not add violation here, then there should be no violation in Records either, because they are basically the same case.

**Comment by romani** [comment]
>I think this is such a rare case, and it does not worth it.

Yes, we can create bug on such cases and wait for next contributor to do next step.
But as you already did a great job to investigate a problem, we can add test cases to our Inputs codebase, and make comments "// violation until #yournewissue"

**Comment by anhminhtran235** [comment]
> But as you already did a great job to investigate a problem, we can add test cases to our Inputs codebase, and make comments "// violation until #yournewissue"

You mean I should add comments like this?
```
$ cat Test.java
public class Test {
  final int x;
  final int y = 1;
  final int z;

  public Test(int x, int y, int z) {
    x = x; // violation
    y = y; // violation until #8973
    z = z; // violation until #8973
  }

  {
    z = 2;
  }
}
```

> Yes, no violation

So, I should disable **all** RequireThis check inside Records?

**Comment by romani** [comment]
> You mean I should add comments like this?
will continue to be violation until #8973 is fixed?

yes. yes

> Maybe the user meant to write x = this.x?

no. this code is nonsence.

> I should disable all RequireThis check inside Records?

please move this point to separate issue, if you can prove that `this` can not be used in records, sure, it is bug.
BUT it is separate bug.

**Comment by anhminhtran235** [comment]
@romani Records in java 15:

```
$ cat Test.java
public record Test(int x, int y) {
  public Test {
    this.x = 2; // compile error
  }

  public Test(int x) {
    this(x, 1);       
    this.x = x; // ok
  }

  public void foo() {   
    this.x = 1; // compile error
  }
}

$ cat confix.xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="RequireThis" />
    </module>
</module>

$ javac --source 15 --enable-preview Test.java
Test.java:3: error: cannot assign a value to final variable x
    this.x = 2;
        ^
Test.java:12: error: cannot assign a value to final variable x
    this.x = 1;
        ^
Note: Test.java uses preview language features.
Note: Recompile with -Xlint:preview for details.
2 errors
```

We almost never can write `this.variable = [value]` since all variables in Records are `private final` by default. I just found out 1 case where we can assign `this.variable` to a value: inside a normal `constructor` (not `compact constructor`).
Since you said `x = this.x` is nonsense, that means I should disable ALL requireThis inside Record EXCEPT when it's inside a normal constructor?

**Comment by anhminhtran235** [comment]
> please move this point to separate issue, if you can prove that this can not be used in records, sure, it is bug.
BUT it is separate bug.

New issue created: https://github.com/checkstyle/checkstyle/issues/8999

**Comment by anhminhtran235** [comment]
> yes. yes

Input added in PR https://github.com/checkstyle/checkstyle/pull/9000
